crates/data-generator/src/blocks/output.rs

```diff
@@ -0,0 +1,628 @@
+                    BlockPropertyVariant::Numeric(n, _) => *n,
+                    _ => unreachable!(),
+                });
+
+                // "AzaleaLeavesDistance::IV"
+                let prop_var_name = prop_var.defused_name().to_ident();
+
+                // "4 => AzaleaLeavesDistance::IV,"
+                let match_arm = quote! { #prop_ord => Self::#prop_var_name, };
+                match_arm.to_tokens(&mut match_arms);
+            }
+
+            let prop_enum_name = self.enum_name;
+
+            tokens.extend(quote! {
+                impl TryFrom<u8> for #prop_enum_name {
+                    type Error = super::UnknownVar;
+
+                    fn try_from(n: u8) -> Result<Self, Self::Error> {
+                        Ok(match n {
+                            #match_arms
+                            _ => Err(super::UnknownVar(n))?,
+                        })
+                    }
+                }
+            });
+        }
+    }
+
+    pub struct PropertyFromBool<'a>(&'a Ident);
+
+    impl<'a> ToTokens for PropertyFromBool<'a> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let prop_enum_name = self.0;
+            tokens.extend(quote! {
+                impl From<bool> for #prop_enum_name {
+                    fn from(value: bool) -> Self {
+                        if value {
+                            Self::True
+                        } else {
+                            Self::False
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    pub struct PropertyIntoBool<'a>(&'a Ident);
+
+    impl<'a> ToTokens for PropertyIntoBool<'a> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let prop_enum_name = self.0;
+            tokens.extend(quote! {
+                impl From<#prop_enum_name> for bool {
+                    fn from(value: #prop_enum_name) -> Self {
+                        value == #prop_enum_name::True
+                    }
+                }
+            });
+        }
+    }
+
+    pub struct PropertyNotBool<'a>(&'a Ident);
+
+    impl<'a> ToTokens for PropertyNotBool<'a> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let prop_enum_name = self.0;
+            tokens.extend(quote! {
+                impl std::ops::Not for #prop_enum_name {
+                    type Output = Self;
+
+                    fn not(self) -> Self::Output {
+                        if self == Self::False {
+                            Self::True
+                        } else {
+                            Self::False
+                        }
+                    }
+                }
+            });
+        }
+    }
+}
+
+trait StringExt {
+    fn to_ident(&self) -> proc_macro2::Ident;
+}
+
+impl<T: AsRef<str>> StringExt for T {
+    fn to_ident(&self) -> proc_macro2::Ident {
+        proc_macro2::Ident::new(self.as_ref(), proc_macro2::Span::call_site())
+    }
+}
```

Comment:
Ideally I'd prefer to have it in a separate module, but as it stands it's not worth another file.